### PR TITLE
Factor out functions for parsers for built-in collections

### LINKED ISSUE #100: Expose built-in parsers for use in custom parsers
In `fgpyo.util.inspect`, the `_get_parser()` function has some useful code for parsing built-in types (set, list, dict, tuple). It would be nice to factor these out into public functions that could be used to implement custom parsers. For example, writing a parser for `collections.Counter` would be as easy as `Counter({key: int(value) for key, value in parse_dict(...).items()})`.

### PR DESCRIPTION
**Pull request by jdidion**
Created public functions for the four built-in collection types handled by `_get_parser`: list, set, tuple, and dict.

Closes #100

### CODE REVIEW

File: fgpyo/util/tests/test_inspect.py [hunk 1]
```diff
@@ -66,3 +74,23 @@ def test_attr_from_custom_type_without_parser_fails() -> None:
...
+
+
+def test_set_parser() -> None:
+    parser = set_parser(Foo, Set[int], {})
+    assert parser("{1,2,3}") == {1, 2, 3}
+
+
+def test_tuple_parser() -> None:
+    parser = tuple_parser(Foo, Tuple[int, str], {})
+    assert parser("(1,a)") == (1, "a")
```
Comment (theJasonFan) [inline fgpyo/util/tests/test_inspect.py:91]: Can we make codecov happy and add `assert parser("()") == ()`? Thanks!

File: fgpyo/util/tests/test_inspect.py [hunk 2]
```diff
@@ -66,3 +74,23 @@ def test_attr_from_custom_type_without_parser_fails() -> None:
...
+
+
+def test_list_parser() -> None:
+    parser = list_parser(Foo, List[int], {})
+    assert parser("1,2,3") == [1, 2, 3]
+
+
+def test_set_parser() -> None:
+    parser = set_parser(Foo, Set[int], {})
+    assert parser("{1,2,3}") == {1, 2, 3}
```
Comment (theJasonFan) [inline fgpyo/util/tests/test_inspect.py:86]: Could also check to parse `{}`, `{1, 1, 2, 3}`.

File: fgpyo/util/tests/test_inspect.py [hunk 3]
```diff
@@ -66,3 +74,23 @@ def test_attr_from_custom_type_without_parser_fails() -> None:
...
+
+
+def test_tuple_parser() -> None:
+    parser = tuple_parser(Foo, Tuple[int, str], {})
+    assert parser("(1,a)") == (1, "a")
+
+
+def test_dict_parser() -> None:
+    parser = dict_parser(Foo, Dict[int, str], {})
+    assert parser("{123;a}") == {123: "a"}
```
Comment (theJasonFan) [inline fgpyo/util/tests/test_inspect.py:96]: Could also check to parse `{}`, and `{123: "a", 123: "b"}`

**Review by theJasonFan — APPROVED** [review]
Looks good to me! Just made some suggestions to tests make code-cov happy and to make edge cases explicitly exampled in tests.

File: fgpyo/util/tests/test_inspect.py (hunk 3 above)
Comment (jdidion) [reply fgpyo/util/tests/test_inspect.py:101]: Here for the second case I did add an explicit check for duplicate keys

File: fgpyo/util/inspect.py [hunk 4]
```diff
@@ -70,6 +70,162 @@ def split_at_given_level(
...
+            out_dict = {}
+            for outer_split in outer_splits:
+                inner_splits = split_at_given_level(outer_split, split_delim=";")
+                assert (
+                    len(inner_splits) % 2 == 0
+                ), "Inner splits of dict didn't have matched key val pairs"
+                for i in range(0, len(inner_splits), 2):
+                    key = key_parser(inner_splits[i])
+                    if key in out_dict:
+                        raise ValueError("Duplicate key found in dict: {}".format(key))
```
Comment (theJasonFan) [inline fgpyo/util/inspect.py:222]: 👍 Nice. I like raising the error here.

File: fgpyo/util/tests/test_inspect.py [hunk 5]
```diff
@@ -66,3 +74,34 @@ def test_attr_from_custom_type_without_parser_fails() -> None:
...
+
+def test_dict_parser() -> None:
+    parser = dict_parser(Foo, Dict[int, str], {})
+    assert parser("{}") == {}
+    assert parser("{123;a}") == {123: "a"}
+
+
+def test_dict_parser_with_duplicate_keys() -> None:
+    parser = dict_parser(Foo, Dict[int, str], {})
+    with pytest.raises(ValueError):
```
Comment (theJasonFan) [inline fgpyo/util/tests/test_inspect.py:106]: 👍

**Review by theJasonFan — APPROVED** [review]
